docs/design-documents/20221024-stream-inspector.md

```diff
@@ -0,0 +1,90 @@
+  inspection data.
+  * A client would need to store the "position" of the previous inspection, to get newer data.
+  * This would require inspection state management in Conduit for each session.
+  * Inspection data would come in with a delay.
+* In a **push based** implementation, Conduit would be pushing the inspection data to a client (HTTP client, gRPC
+  client, UI).
+  * A client would not need to store the "position" of the previous inspection, to get newer data.
+  * Inspection state management in Conduit would be minimal.
+  * Inspection data would come in almost real-time.
+  * Inspecting the data may require additional tools (depends on the actual implementation)
+
+From what we know so far, the push based approach has more advantages and is easier to work with, so that's the 
+approach chosen here. Concrete implementation options are discussed below.
+
+For context: gRPC is the main API for Conduit. The HTTP API is generated using [grpc-gateway](https://github.com/grpc-ecosystem/grpc-gateway).
+
+### API
+For any given pipeline component, only one method is required, one which starts an inspection (i.e. sending of data to 
+the client). 
+
+Three methods are required in total:
+1. One to inspect a connector (there's a single endpoint for source and destination connectors)
+2. One to inspect records coming into a processor
+3. One to inspect records coming out of a processor
+
+### Implementation option 1: WebSockets
+Conduit provides a streaming endpoint. The stream is exposed using the WebSockets API.
+
+grpc-gateway doesn't support WebSockets (see [this](https://github.com/grpc-ecosystem/grpc-gateway/issues/168)). There's 
+an open-source proxy for it though, available [here](https://github.com/tmc/grpc-websocket-proxy/). 
+
+### Implementation option 2: Pure gRPC
+Conduit provides a streaming endpoint. The stream is consumed as such, i.e. a gRPC endpoint. A JavaScript implementation
+of gRPC for browser clients exists (called [grpc-web](https://github.com/grpc/grpc-web)). While that would mean no 
+changes in Conduit itself, it would require a lot of changes in the UI.
+
+### Implementation option 3: Server-sent events 
+TBD
+
+### Chosen implementation
+[grpc-websocket-proxy](https://github.com/tmc/grpc-websocket-proxy/) mention in option 1 is relatively popular and is
+open-source, so using it is no risk. The other option is much costlier.
+
+
+
+## Questions
+
+* Should we rename it to pipeline inspector instead? Pipeline is the term we use in the API, streams are used
+  internally.
+    * Cons: We cannot call it CSI (Conduit's Stream Inspector) anymore.
+* Is metadata needed (such as time the data was captured)?
+* Should there be a limit on how long a stream inspector can run?
+* Should there be a limit on how many records a stream inspector can receive?
+* Are we interested in more than the records? Is there some other data we'd like to see (now or in future)?
```

Comment:
I believe that the records are the main thing we're interested in, so it doesn't look like we should be concerned with other types of data. At the moment, I can't think of something really useful, except the records themselves.

Reply:
Yeah I think that's the main point, we want to be able to inspect the records flowing through the pipeline, no need to design for anything more.

Reply:
Hm some metadata that maybe, could be useful is to show the next node destination? This will help hint at directionality of the message.

Reply:
It looks like this question is a duplicate of the one above "is metadata needed?". 

Regarding directionality of the message: to my knowledge, a message is either dropped (in a processor) or it goes to the next node as specified in the pipeline topology (e.g. another processor, a destination with or without processors etc.). Unless I'm maybe missing something, I'm not sure how much additional value would we actually get from that. Maybe we could do it only to make information more easily accessible (i.e. if it's a complex pipeline, you don't have to go back to the UI to check the topology).

But that reminds of something I think someone already mentioned, and that is adding the information about the nodes through which a record passed, before coming to the inspection point. That might be a useful thing maybe? I believe it would always boil down to the source node.

Reply:
> I believe it would always boil down to the source node.

That's right, if you know the source node you know the path the message has taken to the inspection node. You'd have to know the topology of the pipeline, but that's static and doesn't change while a pipeline is running.

This sounds valuable, I propose we automatically add a metadata field to each record with the source ID or name.

Sidenote: we already have similar fields prepared for metadata about the _connector plugin_ ([here](https://github.com/ConduitIO/conduit-connector-protocol/blob/4b89a6c94401924eb11a2554298be0e5ab3b9b88/proto/connector/v1/connector.proto#L8-L9)) but IIRC they are not used yet.

Reply:
That metadata field is really convenient. :+1: